qiskit/circuit/__init__.py

```diff
@@ -54,81 +78,825 @@
+amount specific to Qiskit.  This is an alphabetical listing of some of the important concepts as a
+quick reference, but does not go into detail of the foundational concepts.  Consider using the `IBM
+Quantum Learning platform <https://learning.quantum.ibm.com/>`_ if you want to start from the
+beginning.
+
+abstract circuit
+    A *circuit* defined in terms of abstract mathematical operations and *virtual qubits*.  This is
+    typically how you think about quantum algorithms; an abstract circuit can be made up of
+    completely arbitrary unitary operations, measurements, and potentially *near-time classical
+    computation*, with no restrictions about which qubits can interact with each other.
+
+    You turn an abstract circuit into a *physical circuit* by using :ref:`Qiskit's transpilation
+    package <qiskit-transpiler>`.
+
+circuit
+    A computational routine the defines a single execution to be taken on a QPU.  This can either be
+    an *abstract circuit* or a *physical circuit*.
+
+gate
+    A *unitary operation* on one or more qubits.
+
+hardware qubit
+    The representation of a single qubit on a particular *QPU*.  A hardware qubit has some physical
+    quantum-mechanical system backing it; unlike a *virtual qubit*, it has particular coupling
+    constraints and only certain gates can be applied to certain physical qubits.
+
+    Qiskit does not distinguish *logical qubits* from any individual *physical qubits* when talking
+    about hardware qubits.  A QPU may implement its hardware qubits as logical qubits, where each
+    hardware qubit comprises many physical qubits that are controlled and error-corrected opaquely
+    to Qiskit by the control electronics.  More likely, for near-term applications, a QPU will be
+    directly exposing its physical qubits as the hardware qubits for Qiskit to reason about.
+
+    Both physical and logical qubits will have coupling constraints between them, only permit
+    certain quantum operations on them, and have scheduling concerns between them.  Qiskit abstracts
+    these concerns together in the concept of hardware qubits. In the early days of quantum error
+    correction, particular backends may let you access their qubit resources either as high-level
+    logical qubits or as low-level physical qubits through Qiskit.
+
+instruction set architecture (ISA)
+    The abstract model of which operations are available on which sets of *hardware qubits* on one
+    particular *QPU*.  For example, one QPU may allow :math:`\sqrt X` and :math:`R_Z` operations on
+    all single hardware qubits, and :math:`CX` operations on certain pairs of hardware qubits.
+
+logical qubit
+    A collection of several *physical qubits* that are controlled together by a QPU (from the user's
+    perspective) to apply real-time quantum error correction.  A logical qubit is a type of
+    *hardware qubit* for Qiskit.
```

Comment:
I think we should not talk about logical qubits at all.

Reply:
Why? I think it's very important to tell people the level of abstraction that Qiskit treats them at, because otherwise it seems like Qiskit won't be performing any abstraction over them.

Reply:
I think the discussion on logical qubits can be left out here because Qiskit does not really process data on that level of abstraction, right? I also think that the concept of logical qubits can be quite orthogonal to the concept of physical and virtual qubits introduced here. For me, logical qubits (or: error-corrected quantum circuits) are defined by their encoding/decoding/gate operation protocol. For these aspects, the qubits can be physical or virtual.

I'm worried that users that know about error-correction will wrongfully assume error correction can be processed in Qiskit natively or there exist abstractions between logical, physical and virtual qubits when in fact, Qiskit does not care about what the qubit do or how they are realized in the end... Maybe the discussion around logical qubits should be reduced to this statement. :-)

User that do not know about error correction will likely just learn some details that they won't be able to apply or control when using Qiskit.

Reply:
> I think the discussion on logical qubits can be left out here because Qiskit does not really process data on that level of abstraction, right? I also think that the concept of logical qubits can be quite orthogonal to the concept of physical and virtual qubits introduced here. For me, logical qubits (or: error-corrected quantum circuits) are defined by their encoding/decoding/gate operation protocol. For these aspects, the qubits can be physical or virtual.

Right, this is the absolute crux of why I think they _should_ be mentioned.  Your paragraph says effectively "you could do either", but in the model I'm describing here, it doesn't make sense to say that a logical qubit a virtual; a virtual qubit is something that exists in Qiskit _entirely_ separated from hardware - they only exist in abstract circuits.

The text you highlight here is largely a formalism of this statement:

> Qiskit does not care about what the [logical] qubit do or how they are realized in the end

and I think it's important to make that clear.  I also mentioned in the surrounding text that hardware vendors might choose to expose both a logical- and a physical-qubit target in the intermediate periods, so Qiskit _could_ be used for manual implementation of EC protocols - both fit within our abstraction of "hardware qubits".